libflash/optimized_writer.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include "optimized_writer.hpp"
+#include <iostream>
+
+using namespace mender;
+
+OptimizedWriter::OptimizedWriter(
+	io::FileReader &reader, io::FileReadWriterSeeker &writer, size_t blockSize, size_t volumeSize) :
+	mBlockSize(blockSize),
+	mReader(reader),
+	mReadWriter(writer),
+	mVolumeSize(volumeSize) {
+}
+
+Error OptimizedWriter::Copy(bool optimized) {
+	mStatistics.mBlocksWritten = 0;
+	mStatistics.mBlocksOmitted = 0;
+	mStatistics.mBytesWritten = 0;
+
+	io::Bytes rv;
+	rv.resize(mBlockSize);
+	io::Bytes wv;
+	wv.resize(mBlockSize);
+
+	while (true) {
+		if (rv.size() != mBlockSize) {
+			rv.resize(mBlockSize);
+		}
+
+		auto pos = mReader.Tell();
+		if (!pos) {
+			return pos.error();
+		}
+		auto position = pos.value();
+
+		if (mVolumeSize && ((position + mBlockSize) > mVolumeSize)) {
+			return NoError;
+		}
```

Comment:
I agree but only if we don't expect scenarios where size of the input is larger that the output volume size (for stdin input we should expect EOF).
If we don't expect that kind of scenarios then it means the volumeSize_ (-s) parameter is only used for error checking because in all valid cases the size of the output volume should be equal to the size of the input?

Reply:
That's correct, they must be equal. The reason for having it is not to do error checking, but because UBI requires knowing the size upfront, and because we will be streaming updates, the total size can not be deduced from the file handle. So we need a separate option for it.